File: kotlin-electron/karakum/src/jsMain/resources/inheritanceModifiers/modifyPropertyInheritance.js
import ts from "typescript"

export default (node) => {
    if (
        ts.isPropertyDeclaration(node)
        && ts.isIdentifier(node.name)
        && (
            (
                (
                    node.name.text === "accessibleTitle"
                    || node.name.text === "autoHideMenuBar"
                    || node.name.text === "closable"
                    || node.name.text === "documentEdited"
                    || node.name.text === "excludedFromShownWindowsMenu"
                    || node.name.text === "focusable"
                    || node.name.text === "fullScreen"
                    || node.name.text === "fullScreenable"
                    || node.name.text === "id"
                    || node.name.text === "kiosk"
                    || node.name.text === "maximizable"
                    || node.name.text === "menuBarVisible"
                    || node.name.text === "minimizable"
                    || node.name.text === "movable"
                    || node.name.text === "representedFilename"
                    || node.name.text === "resizable"
                    || node.name.text === "shadow"
                    || node.name.text === "simpleFullScreen"
                    || node.name.text === "tabbingIdentifier"
                    || node.name.text === "title"
                    || node.name.text === "visibleOnAllWorkspaces"
                )
                && node.parent
                && node.parent.name?.text === "BaseWindow"
            )
        )
    ) {
        return "open"
    }

    if (
        ts.isPropertyDeclaration(node)
        && ts.isIdentifier(node.name)
        && (
            (
                (
                    node.name.text === "accessibleTitle"
                    || node.name.text === "autoHideMenuBar"
                    || node.name.text === "closable"
                    || node.name.text === "documentEdited"
                    || node.name.text === "excludedFromShownWindowsMenu"
                    || node.name.text === "focusable"
                    || node.name.text === "fullScreen"
                    || node.name.text === "fullScreenable"
                    || node.name.text === "id"
                    || node.name.text === "kiosk"
                    || node.name.text === "maximizable"
                    || node.name.text === "menuBarVisible"
                    || node.name.text === "minimizable"
                    || node.name.text === "movable"
                    || node.name.text === "representedFilename"
                    || node.name.text === "resizable"
                    || node.name.text === "shadow"
                    || node.name.text === "simpleFullScreen"
                    || node.name.text === "tabbingIdentifier"
                    || node.name.text === "title"
                    || node.name.text === "visibleOnAllWorkspaces"
                )
                && node.parent
                && node.parent.name?.text === "BrowserWindow"
            )
        )
    ) {
        return "override"
    }

    return null
}

File: kotlin-electron/karakum/src/jsMain/kotlin/electron/karakum/inheritanceModifiers/modifyPropertyInheritance.kt
package electron.karakum.inheritanceModifiers

import electron.karakum.util.nullable
import io.github.sgrishchenko.karakum.extension.InheritanceModifierContext
import io.github.sgrishchenko.karakum.util.getParentOrNull
import typescript.Node
import typescript.isClassDeclaration
import typescript.isIdentifier
import typescript.isPropertyDeclaration

fun modifyPropertyInheritance(node: Node, context: InheritanceModifierContext) = nullable {
    nullable {
        ensure(isPropertyDeclaration(node))

        val name = node.name
        ensure(isIdentifier(name))

        nullable {
            ensure(
                name.text == "accessibleTitle"
                        || name.text == "autoHideMenuBar"
                        || name.text == "closable"
                        || name.text == "documentEdited"
                        || name.text == "excludedFromShownWindowsMenu"
                        || name.text == "focusable"
                        || name.text == "fullScreen"
                        || name.text == "fullScreenable"
                        || name.text == "id"
                        || name.text == "kiosk"
                        || name.text == "maximizable"
                        || name.text == "menuBarVisible"
                        || name.text == "minimizable"
                        || name.text == "movable"
                        || name.text == "representedFilename"
                        || name.text == "resizable"
                        || name.text == "shadow"
                        || name.text == "simpleFullScreen"
                        || name.text == "tabbingIdentifier"
                        || name.text == "title"
                        || name.text == "visibleOnAllWorkspaces"
            )

            val classNode = ensureNotNull(node.getParentOrNull())
            ensure(isClassDeclaration(classNode))
            ensure(classNode.name?.text == "BaseWindow")

            "open"
        } ?: nullable {
            ensure(
                name.text == "accessibleTitle"
                        || name.text == "autoHideMenuBar"
                        || name.text == "closable"
                        || name.text == "documentEdited"
                        || name.text == "excludedFromShownWindowsMenu"
                        || name.text == "focusable"
                        || name.text == "fullScreen"
                        || name.text == "fullScreenable"
                        || name.text == "id"
                        || name.text == "kiosk"
                        || name.text == "maximizable"
                        || name.text == "menuBarVisible"
                        || name.text == "minimizable"
                        || name.text == "movable"
                        || name.text == "representedFilename"
                        || name.text == "resizable"
                        || name.text == "shadow"
                        || name.text == "simpleFullScreen"
                        || name.text == "tabbingIdentifier"
                        || name.text == "title"
                        || name.text == "visibleOnAllWorkspaces"
            )

            val classNode = ensureNotNull(node.getParentOrNull())
            ensure(isClassDeclaration(classNode))
            ensure(classNode.name?.text == "BrowserWindow")

            "override"
        }
    }
}
